commit b25e6bd343299806187911bf6f0cc09efd493f20
Author:     Yuhong Wen <yuhongw@nvidia.com>
AuthorDate: Fri Mar 31 10:18:09 2023 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Fri Mar 31 10:18:09 2023 -0400

    fixed the job status for config error. (#1615)
    
    * fixed the job status for config error.
    
    * Added FINISHED_ABNORMAL state to indicate the job complete with abnormal complete return code.
    
    ---------
    
    Co-authored-by: Chester Chen <512707+chesterxgchen@users.noreply.github.com>
---
 nvflare/private/fed/server/job_runner.py    | 6 +++++-
 nvflare/private/fed/server/server_engine.py | 8 +++++++-
 2 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/nvflare/private/fed/server/job_runner.py b/nvflare/private/fed/server/job_runner.py
index 2affc023..b80ca87b 100644
--- a/nvflare/private/fed/server/job_runner.py
+++ b/nvflare/private/fed/server/job_runner.py
@@ -340,7 +340,11 @@ class JobRunner(FLComponent):
                             )
                             self.abort_client_run(job_id, active_client_sites_names, fl_ctx)
 
-                            job_manager.set_status(job.job_id, RunStatus.FINISHED_EXECUTION_EXCEPTION, fl_ctx)
+                            process_return_code = run_process.get(RunProcessKey.PROCESS_RETURN_CODE)
+                            if process_return_code == -9:
+                                job_manager.set_status(job.job_id, RunStatus.FINISHED_ABNORMAL, fl_ctx)
+                            else:
+                                job_manager.set_status(job.job_id, RunStatus.FINISHED_EXECUTION_EXCEPTION, fl_ctx)
                         else:
                             job_manager.set_status(job.job_id, RunStatus.FINISHED_COMPLETED, fl_ctx)
                         with self.lock:
diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index 68f570ae..b02e89bd 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -223,7 +223,13 @@ class ServerEngine(ServerEngineInternalSpec):
                     self.logger.debug(f"Job:{job_id} UPDATE_RUN_STATUS didn't finish fast enough.")
                     break
                 time.sleep(0.1)
-            self.run_processes.pop(job_id, None)
+            with self.lock:
+                return_code = process.poll()
+                # if process exit but with Execution exception
+                if return_code and return_code != 0:
+                    run_process_info[RunProcessKey.PROCESS_RETURN_CODE] = return_code
+                    self.exception_run_processes[job_id] = run_process_info
+                self.run_processes.pop(job_id, None)
         self.engine_info.status = MachineStatus.STOPPED
 
     def _start_runner_process(
